pkg/migration/migration.go

```diff
@@ -0,0 +1,573 @@
+package migration
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/libopenstorage/operator/drivers/storage"
+	corev1 "github.com/libopenstorage/operator/pkg/apis/core/v1"
+	"github.com/libopenstorage/operator/pkg/constants"
+	"github.com/libopenstorage/operator/pkg/controller/storagecluster"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	portworxDaemonSetName  = "portworx"
+	portworxContainerName  = "portworx"
+	migrationRetryInterval = 30 * time.Second
+	phaseAwaitingApproval  = "AwaitingMigrationApproval"
+)
+
+// Handler object that carries out migration of Portworx Daemonset
+// and it's components to operator managed StorageCluster object
+type Handler struct {
+	client client.Client
+	driver storage.Driver
+}
+
+// New creates a new instance of migration handler
+func New(ctrl *storagecluster.Controller) *Handler {
+	return &Handler{
+		client: ctrl.GetKubernetesClient(),
+		driver: ctrl.Driver,
+	}
+}
+
+// Start starts the migration
+func (h *Handler) Start() {
+	ds, err := h.getPortworxDaemonSet()
+	if err != nil {
+		logrus.Errorf("Migration failed to start. %v", err)
+		return
+	} else if ds == nil {
+		logrus.Infof("Migration is not needed")
+		return
+	}
+
+	cluster, err := h.createStorageClusterIfAbsent(ds)
+	if err != nil {
+		logrus.Errorf("Migration failed to create StorageCluster. %v", err)
+		return
+	}
+
+	wait.PollInfinite(migrationRetryInterval, func() (bool, error) {
+		cluster, isNeeded := h.isMigrationNeeded(cluster)
+		if !isNeeded {
+			return true, nil
+		} else if cluster == nil {
+			return false, nil
+		}
+
+		if !h.isMigrationApproved(cluster) {
```

Comment:
Yes, but we need to keep checking here too, right? This migration handler will start scaling down the daemonset and then delete it once done. We need to start the scale down/up only when the migration is approved.

Reply:
If at this point if migration is not approved yet, this will return false and the handler is done. I am not seeing the handler will be called again after it is approved, did I miss anything?

Reply:
When you return false, the outer retry loop continues. It only exits if you return true or an error.

Reply:
oh yes, thought this is out of the poll loop